### PR TITLE
adds mix tasks paralleling Ecto's for shards and fix main db tasks

### PR DESCRIPTION
**Pull request by drl123**
Tasks added:
- create
- drop
- migrate
- rollback
- load
- dump
- gen.migration

usage: `mix shards.migrate` to migrate all shard db's

Also fixes an issue when trying to migrate the main db's because Ecto creates its own schema for the schema_migrations table that does not inherit from EctoSharding and, as such, does not define the `sharded?` method, causing an exception to be thrown when doing `mix ecto.migrate`.

### CODE REVIEW

**Comment by drl123** [comment]
I don't understand why coverex is claiming that the `run` lines are not tested when I do have tests for them.  I even tried adding tests with different or no arguments and the coverage % did not change.

File: lib/mix/tasks/shards.ex [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  something like `priv/shards` so as not to conflict with the main database migrations.
+"""
+
+  @doc false
+  def set_repo(repo, args) do
+    {parsed, args, other} = OptionParser.parse(args, aliases: [r: :repo])
+
+    permitted = parsed
+                |> Keyword.merge(repo: repo)
+                |> OptionParser.to_argv()
```
Comment (craig-day) [inline lib/mix/tasks/shards.ex]: I don't have credo setup yet, but I prefer the pipeline-on-a-new-line approach

```elixir
permitted =
  parsed
  |> Keyword.merge(repo: repo)
  |> OptionParser.to_argv
```

**Comment by craig-day** [comment]
This is amazing! I've started reviewing and hopefully will have an initial review done in the next few days. I wouldn't worry about coverex for now, we can figure out why its being dumb later.

**Comment by drl123** [comment]
Just remember I'm an Elixir Noob!  ;-)

If you've got any words of wisdom on better structure, I'm all ears.  I'm sure we could probably DRY up the tasks a bit, but then again, they only get used in this one file, so I felt the explicitness was beneficial here.

File: README.md [hunk 2]
```diff
@@ -51,6 +52,9 @@ end
     ```
 
     If you know your shard information at compile time, you can also add that.
+    
+    Note: Be sure to specify your path where you want the shard migrations to go with the `priv` key.  In the example
```
Comment (craig-day) [inline README.md]: Nit: Can we clear up the wording? Maybe something like 

> Configure the path to the shard migration files using the `migrations_path` configuration key.

File: README.md [hunk 3]
```diff
@@ -62,15 +66,17 @@ end
           password: System.get_env("MYSQL_PASSWORD"),
           database: "my_db_shard_1",
           hostname: "10.0.0.1",
-          pool_size: 15
+          pool_size: 15,
+          priv: "priv/shards"
```
Comment (craig-day) [inline README.md:73]: Can we make this key a little more explicit? Maybe `migrations_path`?

File: lib/mix/tasks/shards.ex (hunk 1 above)
Comment (craig-day) [reply lib/mix/tasks/shards.ex]: also, functions in a pipeline that only accept the one argument don't need empty parens

File: lib/mix/tasks/shards.ex [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      |> List.first()
+      |> Shards.set_repo(args)
+      |> Mix.Tasks.Ecto.Gen.Migration.run()
+    end
+  end
+
+  defmodule Dump do
+    use Mix.Task
+    alias Mix.Tasks.Shards
+    alias EctoSharding.Configuration, as: Config
```
Comment (craig-day) [inline lib/mix/tasks/shards.ex]: a couple thoughts on this pattern. 

1. I'm not crazy about having nested `defmodule`s in here
1. These same three lines are in every module

to address these, what do you think about this

1. can we put each migration action in its own file, similar to [what Ecto does](https://github.com/elixir-ecto/ecto/tree/master/lib/mix/tasks)
1. Maybe the base module, `Mix.Tasks.Shards`, can define a `__using__` macro which does these three things

File: lib/mix/tasks/shards.ex [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    @doc "Loads the Shard DB's from the schema dump file"
+    def run(args \\ []) do
+      Enum.each(Config.shard_repos(), fn({_name, repo}) ->
+        Shards.set_repo(repo, args)
+        |> Mix.Tasks.Ecto.Load.run()
+      end)
```
Comment (craig-day) [inline lib/mix/tasks/shards.ex]: I see two basic patterns, here: 

1. we iterate over all the shards and call `Mix.Tasks.Ecto.XXX.run()`
1. we grab the first shard and call `Mix.Tasks.Ecto.XXX.run()`

thoughts on cleaning this up a little bit? Could maybe extract a few methods to help with these? I'm happy to push up a commit to work on this, but I probably won't get to it until the weekend.

File: lib/ecto_sharding/query_processing.ex [hunk 6]
```diff
@@ -30,7 +30,7 @@ defmodule EctoSharding.QueryProcessing do
           schema -> schema
         end
 
-      if model.sharded? do
+      if Keyword.has_key?(model.__info__(:functions), :sharded?) && model.sharded? do
```
Comment (craig-day) [inline lib/ecto_sharding/query_processing.ex:33]: why did this get added again? would it be worth pulling this change into a separate PR to keep concerns separate here, or was this specific to the migrations work?

**Review by craig-day — CHANGES_REQUESTED** [review]
This looks great! The approach seems solid, just a few opportunities to clean things up.

FYI: I'm probably going to push a commit to master to remove the docs directory to make changes like this much cleaner. I'll let you know and you can rebase to remove a bunch of these files.

File: README.md (hunk 3 above)
Comment (drl123) [reply README.md:73]: The `priv` key comes from Ecto.  I would prefer to maintain Ecto's naming to avoid confusion.  The `priv` directory also holds the seeds and schema files in addition to the migrations folder.

File: README.md (hunk 2 above)
Comment (drl123) [reply README.md]: yep

File: lib/mix/tasks/shards.ex (hunk 1 above)
Comment (drl123) [reply lib/mix/tasks/shards.ex]: easy enough...I'll do that throughout.

File: lib/ecto_sharding/query_processing.ex (hunk 6 above)
Comment (drl123) [reply lib/ecto_sharding/query_processing.ex:33]: This makes migrations on the main db possible because when Ecto tries to read the schema_migrations it fails because it uses Ecto.Repo.

Comment (drl123) [reply lib/ecto_sharding/query_processing.ex:33]: Sorry, wasn't Ecto.Repo....was the fact that there is no Schema defined in the project for schema_migrations.  Ecto builds that on the fly and there's no opportunity to make it inherit from EctoSharding.Schema, so the sharded? method doesn't exist and blows up.  Changing this allows it to work when the actual 'model' is an EctoSharding.Schema, but the schema migrations is not.  I don't think there is any equivalent of a `try` method in Elixir, is there?  That's effectively what this is doing.

File: lib/mix/tasks/shards.ex (hunk 5 above)
Comment (drl123) [reply lib/mix/tasks/shards.ex]: Any suggestions on the tests?  I'm thinking maybe I need to define another set of Repos in the test config for these tests to operate on.  I think the random failures are due to the database being modified, which is not in a transaction, while the other tests are running, so part way through their db evaporates!

**Comment by drl123** [comment]
Sorry, forgot you had Travis hooked up and it was firing off builds with each push.  I split the tasks into their own files by module, now just have to DRY things up, fix the doc's and do something about the test situation.  My module tests run fine in isolation, but hoses the rest of the tests randomly.  Hmm...should probably separate them into individual test files too, to mirror the file structure of the code.

I'll try to get some time on the weekend to tidy that stuff up.

**Comment by drl123** [comment]
I think I got all of the changes you requested covered.  All we need to deal with now are the pesky tests.

**Comment by drl123** [comment]
So did some testing on my local machine and the Master branch has random test failures too.  Since my test should be mocking out the Ecto task and not calling the real implementation, I don't think they are causing the random build failures.  You may need to set your tests to async: false...just a hunch.

**Comment by drl123** [comment]
OK...I give up on Coveralls.  I thought since it put the coverage at 50% on each method, it was complaining about arity coverage, so I added tests with arity of 0 and 1, but no change in %.  I did see some posts from the Elixir team where coveralls was flaking out when there were defaults specified in the arguments.  Maybe we're suffering the same problem?

https://github.com/elixir-lang/elixir/issues/3847

**Comment by drl123** [comment]
FYI - I tried separating the run function into one of the tasks to run/0 and run/1 to see if it changed anything, but still same.  So the default args are not the issue.

**Comment by craig-day** [comment]
hey @drl123 I haven't forgotten about this. I will try my best to look today or tomorrow. Thanks for all the work!

**Comment by craig-day** [comment]
@drl123 if you rebase with master, you should be able to remove the docs directory altogether.

**Review by craig-day — APPROVED** [review]
I've delayed this for too long. I'm going to merge as I assume it works for your cases. I'll work on cleaning up any of the nit picky stuff I have about styling.

**Comment by craig-day** [comment]
@drl123 this is published in v0.0.8

**Comment by drl123** [comment]
Thank you!
